File: assign1/Scala/src/test/scala/NameTest.scala
package assign1

import org.scalatest.FunSuite
import assign1.Name

class NameTest extends FunSuite {

  test("canary"){
    assert(true)
  }

  test("imperativeAverageWithEmptyList"){
    assert(Name.imperativeAverageNumberOfLetters(List()) == 0)
  }

  test("imperativeAverageWithOneNameInList"){
    assert(Name.imperativeAverageNumberOfLetters(List("Derrick")) == 7)
  }

  test("imperativeAverageWithTwoNamesOfSameLengthInList"){
    assert(Name.imperativeAverageNumberOfLetters(List("Omar", "Hiep")) == 4)
  }

  test("imperativeAverageWithTwoNamesOfDifferentLengthInList"){
    assert(Name.imperativeAverageNumberOfLetters(List("Derrick", "Al")) == 4.5)
  }

  test("imperativeAverageWithThreeNamesOfDifferentLengthInList"){
    assert(Name.imperativeAverageNumberOfLetters(List("Derrick", "Alexander", "Ira")) == 19.0/3.0)
  }

  test("imperativeGroupNamesByFirstLetterWithEmptyList"){
    assert(Name.imperativeGroupNamesByFirstLetter(List()) == Map())
  }

  test("imperativeGroupNamesByFirstLetterWithOneName"){
    assert(Name.imperativeGroupNamesByFirstLetter(List("Jim")) == Map('J' -> List("Jim")))
  }

  test("imperativeGroupNamesByFirstLetterWithTwoNamesWithSameStartingLetter"){
    assert(Name.imperativeGroupNamesByFirstLetter(List("Jim", "Jack")) == Map('J' -> List("Jim", "Jack")))
  }

  test("imperativeGroupNamesByFirstLetterWithThreeNamesWithDifferentStartingLetters"){
    assert(Name.imperativeGroupNamesByFirstLetter(List("Jim", "Jack", "Sara")) ==
      Map('J' -> List("Jim", "Jack"), 'S' -> List("Sara")))
  }

  test("imperativeCountNamesStartingWithEachLetterWithEmptyList"){
    assert(Name.imperativeCountNamesStartingWithEachLetter(List()) == Map())
  }

  test("imperativeCountNamesStartingWithEachLetterWithOneName"){
    assert(Name.imperativeCountNamesStartingWithEachLetter(List("Jim")) == Map('J' -> 1))
  }

  test("imperativeCountNamesStartingWithEachLetterWithTwoNamesWithSameStartingLetter"){
    assert(Name.imperativeCountNamesStartingWithEachLetter(List("Jim", "Jack")) == Map('J' -> 2))
  }

  test("imperativeCountNamesStartingWithEachLetterWithThreeNamesWithDifferentStartingLetters"){
    assert(Name.imperativeCountNamesStartingWithEachLetter(List("Jim", "Jack", "Sara")) == Map('J' -> 2, 'S' -> 1))
  }

  test("functionalAverageWithEmptyList"){
    assert(Name.functionalAverageNumberOfLettersInName(List()) == 0.0)
  }

  test("functionalAverageWithOneNameInList"){
    assert(Name.functionalAverageNumberOfLettersInName(List("Derrick")) == 7.0)
  }

  test("functionalAverageWithTwoNamesOfSameLengthInList"){
    assert(Name.functionalAverageNumberOfLettersInName(List("Omar", "Hiep")) == 4.0)
  }

  test("functionalAverageWithTwoNamesOfDifferentLengthInList"){
    assert(Name.functionalAverageNumberOfLettersInName(List("Derrick", "Alexander")) == 8.0)
  }

  test("functionalAverageWithThreeNamesOfDifferentLengthInList"){
    assert(Name.functionalAverageNumberOfLettersInName(List("Derrick", "Alexander", "Ira")) == 19.0/3.0)
  }

  test("functionalGroupNamesByFirstLetterEmptyList") {
    assert(Name.functionalGroupNamesByFirstLetter(List()) == Map())
  }

  test("functionalGroupNamesByFirstLetterOneName") {
    assert(Name.functionalGroupNamesByFirstLetter(List("Jim")) == Map('J' -> List("Jim")))
  }

  test("functionalGroupNamesByFirstLetterTwoNamesSameLetter") {
    assert(Name.functionalGroupNamesByFirstLetter(List("Jim", "Jack")) == Map('J' -> List("Jim", "Jack")))
  }

  test("functionalGroupNamesByFirstLetterTwoNamesDifferentLetters") {
    assert(Name.functionalGroupNamesByFirstLetter(List("Jim", "Jack", "Sara")) == Map('J' -> List("Jim", "Jack"), 'S' -> List("Sara")))
  }

  test("functionalCountNamesStartingWithEachLetterWithEmptyList"){
    assert(Name.functionalCountNamesStartingWithEachLetter(List()) == Map())
  }

  test("functionalCountNamesStartingWithEachLetterWithOneName"){
    assert(Name.functionalCountNamesStartingWithEachLetter(List("Jim")) == Map('J' -> 1))
  }

  test("functionalCountNamesStartingWithEachLetterWithTwoNamesWithSameStartingLetter"){
    assert(Name.functionalCountNamesStartingWithEachLetter(List("Jim", "Jack")) == Map('J' -> 2))
  }

  test("functionalCountNamesStartingWithEachLetterWithThreeNamesWithDifferentStartingLetters"){
    assert(Name.functionalCountNamesStartingWithEachLetter(List("Jim", "Jack", "Sara")) == Map('J' -> 2, 'S' -> 1))
  }
}

File: assign1/PHP/test/NameTest.php
<?php

include_once 'src/Name.php';

echo "Canary test\n";
assert(true);

echo "IMPERATIVE TESTS\n";
echo "Average number of letters per name, empty list\n";
assert(imperativeAverageNumberOfLetters(array()) == 0.0);
echo "Average number of letters per name, one name in list\n";
assert(imperativeAverageNumberOfLetters(array("Derrick")) == 7.0);
echo "Average number of letters per name, two names in list, same length\n";
assert(imperativeAverageNumberOfLetters(array("Omar", "Hiep")) == 4.0);
echo "Average number of letters per name, two names in list, different length\n";
assert(imperativeAverageNumberOfLetters(array("Derrick", "Al")) == 9.0/2.0);
echo "Average number of letters per name, three names in list, different length\n";
assert(imperativeAverageNumberOfLetters(array("Derrick", "Alexander", "Ira")) == 19.0/3.0);
echo "Group Names By First Letter, Empty List\n";
assert(imperativeGroupNamesByFirstLetter(array()) === []);
echo "Group Names By First Letter, One Name\n";
assert(imperativeGroupNamesByFirstLetter(array("Jim")) === array("J" => ["Jim"]));
echo "Group Names By First Letter, Two Names, Same Starting Letter\n";
assert(imperativeGroupNamesByFirstLetter(array("Jim", "Jack")) === array("J" => ["Jim", "Jack"]));
echo "Group Names By First Letter, Three Names, Different Starting Letters\n";
assert(imperativeGroupNamesByFirstLetter(array("Jim", "Jack", "Sara")) === array("J" => ["Jim", "Jack"], "S" => ["Sara"]));
echo "Count Names By First Letter, Empty List\n";
assert(imperativeCountNamesByFirstLetter(array()) === array());
echo "Count Names By First Letter, One Name\n";
assert(imperativeCountNamesByFirstLetter(array("Jim")) === array("J" => 1));
echo "Count Names By First Letter, Two Names, Same Starting Letter\n";
assert(imperativeCountNamesByFirstLetter(array("Jim", "Jack")) === array("J" => 2)); 
echo "Count Names By First Letter, Three Names, Different Starting Letter\n";
assert(imperativeCountNamesByFirstLetter(array("Jim", "Jack", "Sara")) === array("J" => 2, "S" => 1)); 

echo "FUNCTIONAL TESTS\n";
echo "Average number of letters per name, empty list\n";
assert(functionalAverageNumberOfLetters(array()) == 0.0);
echo "Average number of letters per name, one name in list\n";
assert(functionalAverageNumberOfLetters(array("Derrick")) == 7.0);
echo "Average number of letters per name, two names in list, same length\n";
assert(functionalAverageNumberOfLetters(array("Omar", "Hiep")) == 4.0);
echo "Average number of letters per name, two names in list, different length\n";
assert(functionalAverageNumberOfLetters(array("Derrick", "Al")) == 9.0/2.0);
echo "Average number of letters per name, three names in list, different length\n";
assert(functionalAverageNumberOfLetters(array("Derrick", "Alexander", "Ira")) == 19.0/3.0);
echo "Group Names By First Letter, Empty List\n";
assert(functionalGroupNamesByFirstLetter(array()) === []);
echo "Group Names By First Letter, One Name\n";
assert(functionalGroupNamesByFirstLetter(array("Jim")) === array("J" => ["Jim"]));
echo "Group Names By First Letter, Two Names, Same Starting Letter\n";
assert(functionalGroupNamesByFirstLetter(array("Jim", "Jack")) === array("J" => ["Jim", "Jack"]));
echo "Group Names By First Letter, Three Names, Different Starting Letters\n";
assert(functionalGroupNamesByFirstLetter(array("Jim", "Jack", "Sara")) === array("J" => ["Jim", "Jack"], "S" => ["Sara"]));
echo "Count Names By First Letter, Empty List\n";
assert(functionalCountNamesByFirstLetter(array()) === array());
echo "Count Names By First Letter, One Name\n";
assert(functionalCountNamesByFirstLetter(array("Jim")) === array("J" => 1));
echo "Count Names By First Letter, Two Names, Same Starting Letter\n";
assert(functionalCountNamesByFirstLetter(array("Jim", "Jack")) === array("J" => 2));
echo "Count Names By First Letter, Three Names, Different Starting Letter\n";
assert(functionalCountNamesByFirstLetter(array("Jim", "Jack", "Sara")) === array("J" => 2, "S" => 1));

echo "Testing complete\n"; 
?>
